File: 5-Problem-A-Day/61. Rotate List.java
/**
 * Definition for singly-linked list. public class ListNode { int val; ListNode next; ListNode(int
 * x) { val = x; } }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if (head == null)
            return null;
        ListNode last = head;
        int size = 1;
        while (last.next != null) {
            size += 1;
            last = last.next;
        }
        k %= size;
        ListNode curr = head;
        int count = 1;
        while (curr != null && curr.next != null && count < size - k) {
            curr = curr.next;
            count += 1;
        }
        last.next = head;
        head = curr.next;
        curr.next = null;
        return head;
    }
}

File: 2021/61. Rotate List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if (head == NULL || k == 0) return head;
        int len = 0;
        ListNode *curr = head, *last = head;
        while (curr != NULL) {
            last = curr;
            curr = curr->next;            
            len++;
        }   
        k = k % len;        
        curr = head;
        for (int i = 0; i < len - k - 1; i++) {
            curr = curr->next;
        }        
        ListNode *newHead = curr->next;
        curr->next = NULL;
        if (newHead == NULL) {
            return head;
        }
        last->next = head;
        return newHead;
    }
};


/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if (head == NULL || head->next == NULL || k == 0) return head;
        
        ListNode* oldTail = head;
        int n = 1;
        for (; oldTail->next != NULL; n++) {
            oldTail = oldTail->next;
        }
        oldTail->next = head;
        
        ListNode* newTail = head;
        for (int i = 0; i < n - k % n - 1; i++) {
            newTail = newTail->next;            
        }
        ListNode* newHead = newTail->next;
        
        newTail->next = NULL;
        return newHead;
        
    }
};